crates/sui-network/src/randomness/mod.rs

```diff
@@ -574,18 +580,81 @@ impl RandomnessEventLoop {
         }
 
         debug!("successfully generated randomness full signature");
-        self.completed_sigs.insert((epoch, round));
-        self.remove_partial_sigs_in_range(sig_bounds);
+        self.process_valid_full_signature(epoch, round, sig);
+    }
+
+    #[instrument(level = "debug", skip_all, fields(?peer_id, ?epoch, ?round))]
+    fn receive_full_signature(
+        &mut self,
+        peer_id: PeerId,
+        epoch: EpochId,
+        round: RandomnessRound,
+        sig: RandomnessSignature,
+    ) {
+        let vss_pk = {
+            let Some(dkg_output) = &self.dkg_output else {
+                debug!("called receive_full_signature before DKG completed");
+                return;
+            };
+            &dkg_output.vss_pk
+        };
+
+        // Basic validity checks.
+        if epoch != self.epoch {
+            debug!("skipping received full sig, we are on epoch {}", self.epoch);
+            return;
+        }
+        if self.completed_sigs.contains_key(&round) {
+            debug!("skipping received full sigs, we already have completed this sig");
+            return;
+        }
+        let highest_completed_round = self.highest_completed_round.get(&epoch).copied();
+        if let Some(highest_completed_round) = &highest_completed_round {
+            if *highest_completed_round >= round {
+                debug!("skipping received full sig, we already have completed this round");
+                return;
+            }
+        }
+
+        if let Err(e) =
+            ThresholdBls12381MinSig::verify(vss_pk.c0(), &round.signature_message(), &sig)
+        {
+            info!("received invalid full signature from peer {peer_id}: {e:?}");
+            return;
+        }
+
+        debug!("received valid randomness full signature");
+        self.process_valid_full_signature(epoch, round, sig);
+    }
+
+    fn process_valid_full_signature(
+        &mut self,
+        epoch: EpochId,
+        round: RandomnessRound,
+        sig: RandomnessSignature,
+    ) {
+        assert_eq!(epoch, self.epoch);
+
+        if let Some((_, full_sig_cell)) = self.send_tasks.get(&round) {
```

Comment:
Don't we process those requests concurrently? if so, two requests with full signatures might trigger code more than once since we update completed_sigs below, no?